### PR TITLE
feat: Add GET Route /defi/rewards to Centralize DeFi Spring Reward Collection

### PR DESCRIPTION
**Pull request by JoE11-y**
### Work Done

1. Route and Input Validation: Implemented /defi/rewards route to centralize reward collection, with addr parameter validation.
2. HTTP Client with Middleware: Configured retry policy and tracing middleware for resilient, traceable HTTP requests.
3. Protocol Data Fetching: Created fetch rewards functions for ZkLend, Nostra, Nimbora, and Ekubo, handling errors by defaulting to empty vectors.
4. Centralized Reward Struct: Defined CommonReward struct and RewardSource enum to standardize protocol data.
5. Call Generation: Implemented create_calls to map reward data to Call objects for streamlined claiming.
6. JSON Response Construction: Built a unified response with rewards and calls, grouped by protocol for easy consumption.
7. Error Handling
8. Integration

**N/B**: Due to the change in `reqwest` version to support the middleware, I updated all `StatusCode` references to use Axum's `StatusCode` instead.

### Call Data

The data structures for each protocol claims are outlined below, along with links to sample transactions that informed the structure.

#### 1. Nimbora
- **Sample Transaction**: [Nimbora Claim on StarkScan](https://starkscan.co/call/0x06da50f3a0fb3f96032afed3ab50e567f2bbb250d17029d104c54723917055a8_1_1)
- **Data Structure**:
  ```rust
  {
      amount: String,
      proof: Vec<String>
  }
  ```

#### 2. ZkLend
- **Sample Transaction**: [ZkLend Claim on StarkScan](https://starkscan.co/call/0x04399e0892d054c744eb768123b72fba98200779574a403926d8133c7040907c_1_7)
- **Data Structure**:
  ```rust
  {
      claim: {
          id: u64,
          amount: String,
          claimee: String,
      },
      proof: Vec<String>
  }
  ```

#### 3. Nostra
- **Sample Transaction**: [Nostra Claim on StarkScan](https://starkscan.co/call/0x02ae9e6a0abd0be6fac315a20d16a35240f58d70f777f5532cdc02e6734b5379_1_25)
- **Data Structure**:
  ```rust
  {
      amount: String,
      proof: Vec<String>
  }
  ```

#### 4. Ekubo
- **Sample Transaction**: [Ekubo Claim on StarkScan](https://starkscan.co/call/0x03e6562b720cd7fddc8503550c98569cd85f44ee19dfefe0a0fa7b5bc8b5b235_1_1)
- **Data Structure**:
  ```rust
  {
      claim: {
          id: u64,
          amount: String,
          claimee: String,
      },
      proof: Vec<String>
  }
  ```
  
  
  @Marchand-Nicolas Any insights or optimizations would be appreciated.

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1 @@
+HOST="http://0.0.0.0:8080"
```
Comment (Marchand-Nicolas) [inline .env]: Remove this part

File: src/endpoints/defi/rewards.rs [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+use std::sync::Arc;
+
+// Define constants for token and contract addresses
+const STRK_TOKEN: &str = "0x04718f5a0fc34cc1af16a1cdee98ffb20c31f5cd61d6ab07201858f4287c938d";
+const STRK_DECIMALS: u8 = 18;
+const STRK_NAME: &str = "STRK";
+const NIMBORA_CLAIM_CONTRACT: &str =
+    "0x07ed46700bd12bb1ee8a33a8594791003f9710a1ab18edd958aed86a8f82d3d1";
+const NOSTRA_CLAIM_CONTRACT: &str =
+    "0x008faa2edc6833a6ad0625f1128d56cf471c3f9649ff2201d9ef49d7e9bb18de";
```
Comment (Marchand-Nicolas) [inline src/endpoints/defi/rewards.rs]: Set these constants in config.toml & config.template.toml

File: src/endpoints/defi/rewards.rs [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    pub reward_source: RewardSource,
+    pub claimed: bool,
+}
+
+#[route(get, "/defi/rewards")]
+pub async fn get_defi_rewards(
+    State(_): State<Arc<AppState>>,
+    Query(params): Query<HashMap<String, String>>,
+) -> impl IntoResponse {
+    let addr = match params.get("addr") {
```
Comment (Marchand-Nicolas) [inline src/endpoints/defi/rewards.rs]: We often use dec and not hex number, you can check in `src/endpoints/has_completed_quest.rs` how we manage to handle both easily

File: src/endpoints/defi/rewards.rs [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+        .headers(get_headers())
+        .json(&nostra_request_body)
+        .send()
+        .await?;
+    match response.json::<NostraResponse>().await {
+        Ok(result) => {
+            let rewards = result
+                .documents
+                .into_iter()
+                .map(|doc| CommonReward {
```
Comment (Marchand-Nicolas) [inline src/endpoints/defi/rewards.rs]: No need to return all these fields for the rewards array. Only keep the amount and the token name.
The aim is to have the necessary data to build this modal:
image
(Except Vesu that we'll add later)

File: src/endpoints/defi/rewards.rs [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+            Ok(rewards)
+        }
+        Err(err) => {
+            eprintln!("Failed to deserialize ekubo response: {:?}", err);
+            Err(Error::Reqwest(err))
+        }
+    }
+}
+
+fn create_calls(rewards: &[CommonReward], addr: &str) -> Vec<Call> {
```
Comment (Marchand-Nicolas) [inline src/endpoints/defi/rewards.rs]: When testing the endpoint, the calls field is not an array:
image.
This should simply be an array as following:
```json
"calls": [
   {
        "contract": "0x1a9acedc9a858ec3a2b6e5b2540aab67c6fab3376c5d56b44dfc3c18d727663",
        "call_data": [...],
        "entry_point": "claim",
   },
...
]
````
Also, why did you added a "regex" field ?

In addition, the calldata is not well formatted.
image
This should be an array of string. Every calldata should be formatted depending the protocol. To know what format each protocol expects, check on starkscan their contract. E.g. for Zklend:
image

**Review by Marchand-Nicolas — CHANGES_REQUESTED** [review]
Some changes are required to be able to run the multicall.

Also, don't forget to try putting the returned call_data in a multicall to test if it works

File: .env (hunk 1 above)
Comment (JoE11-y) [reply .env]: alright noted.

File: src/endpoints/defi/rewards.rs (hunk 2 above)
Comment (JoE11-y) [reply src/endpoints/defi/rewards.rs]: alright noted.

File: src/endpoints/defi/rewards.rs (hunk 3 above)
Comment (JoE11-y) [reply src/endpoints/defi/rewards.rs]: okay will take a look at it.

File: src/endpoints/defi/rewards.rs (hunk 4 above)
Comment (JoE11-y) [reply src/endpoints/defi/rewards.rs]: alright will remove redundant data.

File: src/endpoints/defi/rewards.rs (hunk 5 above)
Comment (JoE11-y) [reply src/endpoints/defi/rewards.rs]: Okay will crosscheck the calls,

I actually reused the call struct in the models and the regex was a requirement there to be used in endpoints/quests/verify_contract

maybe i'll just create a new struct for it.

Comment (Marchand-Nicolas) [reply src/endpoints/defi/rewards.rs]: Yes it's not for the same thing

**Review by Marchand-Nicolas — APPROVED** [review]
Lgtm!

**Pull request merged**